botocore/useragent.py

```diff
@@ -0,0 +1,421 @@
+
+        For backwards-compatibility both session-level and client-level config
+        of custom tool names are honored. If this removes the Botocore
+        information from the start of the string, Botocore's name and version
+        are included as a separate field with "md" prefix.
+        """
+        if (
+            self._session_user_agent_name
+            and self._session_user_agent_version
+            and (
+                self._session_user_agent_name != _USERAGENT_SDK_NAME
+                or self._session_user_agent_version != botocore_version
+            )
+        ):
+            return [
+                UserAgentComponent(
+                    self._session_user_agent_name,
+                    self._session_user_agent_version,
+                ),
+                UserAgentComponent(
+                    'md', _USERAGENT_SDK_NAME, botocore_version
+                ),
+            ]
+
+        return [UserAgentComponent(_USERAGENT_SDK_NAME, botocore_version)]
+
+    def _build_os_metadata(self):
+        """
+        Build the OS/platform components of the User-Agent header string.
+
+        For recognized platform names that match or map to an entry in the list
+        of standardized OS names, a single component with prefix "os" is
+        returned. Otherwise, one component "os/other" is returned and a second
+        with prefix "md" and the raw platform name.
+
+        String representations of example return values:
+         * ``os/macos#10.13.6``
+         * ``os/linux``
+         * ``os/other``
+         * ``os/other md/foobar#1.2.3``
+         * ``os/linux md/arch#x86_64``
+        """
+        if self._platform_name is None:
+            return [UserAgentComponent('os', 'other')]
+
+        plt_name_lower = self._platform_name.lower()
+        if plt_name_lower in _USERAGENT_ALLOWED_OS_NAMES:
+            os_family = plt_name_lower
+        elif plt_name_lower in _USERAGENT_PLATFORM_NAME_MAPPINGS:
+            os_family = _USERAGENT_PLATFORM_NAME_MAPPINGS[plt_name_lower]
+        else:
+            os_family = None
+
+        if os_family is not None:
+            return [
+                UserAgentComponent('os', os_family, self._platform_version)
+            ]
+        else:
+            return [
+                UserAgentComponent('os', 'other'),
+                UserAgentComponent(
+                    'md', self._platform_name, self._platform_version
+                ),
+            ]
+
+    def _build_language_metadata(self):
+        """
+        Build the language components of the User-Agent header string.
+
+        Returns the Python version in a component with prefix "lang" and name
+        "python". The Python implementation (e.g. CPython, PyPy) is returned as
+        separate metadata component with prefix "md" and name "pyimpl".
+
+        String representation of an example return value:
+        ``lang/python/3.10.4 md/pyimpl/Cpython``
```

Comment:
```suggestion
        ``lang/python/3.10.4 md/pyimpl/CPython``
```

Reply:
I find this choice interesting when you look at Cython, Jython, IronPython, etc where the version may not match a Python version. This brings up a question about how robust this needs to be. Do we have enough encapsulation that the client will never fail during user_agent construction? If so, what's our fallback?

That may be answered below so feel free to ignore (or I should come back and delete this).

Reply:
The only assumption we make is that we get a string back from [platform.python_implementation](https://docs.python.org/3.8/library/platform.html#platform.python_implementation). I believe that can be relied on.